src/interface/blas2_interface.hpp

```diff
@@ -714,6 +777,57 @@ typename sb_handle_t::event_t inline _trmv(sb_handle_t& sb_handle, char _Uplo,
+    if (tolower(_trans) == 'n') {                                 \
+      if (tolower(_Diag) == 'n') {                                \
+        return func<uplo_type::Upper, transpose_type::Normal,     \
+                    diag_type::Nonunit>(__VA_ARGS__);             \
+      } else {                                                    \
+        return func<uplo_type::Upper, transpose_type::Normal,     \
+                    diag_type::Unit>(__VA_ARGS__);                \
+      }                                                           \
+    } else {                                                      \
+      if (tolower(_Diag) == 'n') {                                \
+        return func<uplo_type::Upper, transpose_type::Transposed, \
+                    diag_type::Nonunit>(__VA_ARGS__);             \
+      } else {                                                    \
+        return func<uplo_type::Upper, transpose_type::Transposed, \
+                    diag_type::Unit>(__VA_ARGS__);                \
+      }                                                           \
+    }                                                             \
+  } else {                                                        \
+    if (tolower(_trans) == 'n') {                                 \
+      if (tolower(_Diag) == 'n') {                                \
+        return func<uplo_type::Lower, transpose_type::Normal,     \
+                    diag_type::Nonunit>(__VA_ARGS__);             \
+      } else {                                                    \
+        return func<uplo_type::Lower, transpose_type::Normal,     \
+                    diag_type::Unit>(__VA_ARGS__);                \
+      }                                                           \
+    } else {                                                      \
+      if (tolower(_Diag) == 'n') {                                \
+        return func<uplo_type::Lower, transpose_type::Transposed, \
+                    diag_type::Nonunit>(__VA_ARGS__);             \
+      } else {                                                    \
+        return func<uplo_type::Lower, transpose_type::Transposed, \
+                    diag_type::Unit>(__VA_ARGS__);                \
+      }                                                           \
+    }                                                             \
+  }
+
+template <typename sb_handle_t, typename index_t, typename container_t0,
+          typename container_t1, typename increment_t>
+typename sb_handle_t::event_t inline _trsv(sb_handle_t& sb_handle, char _Uplo,
+                                           char _trans, char _Diag, index_t _N,
+                                           container_t0 _mA, index_t _lda,
+                                           container_t1 _vx,
+                                           increment_t _incx) {
+  INST_UPLO_TRANS_DIAG(blas::trsv::backend::_trsv, sb_handle, _N, _mA, _lda,
```

Comment:
I would more prefer if you could replace macro with C++ code instead

Reply:
Yes, do you have suggestions?

I ended up using a macro for solving the problem from the point of view of text manipulation since I didn't figure out how to deal easily with non instantiated template functions (i.e., `blas::trsv::backend::_trsv`).

Reply:
You can pass a function pointer if you want, but the cleaner solution would be to have a different struct which instantiates each of the function via operator maybe and you pass those struct type as a template for each case and call the operator for it.

Reply:
Since this is a common problem shared among different routines, we should come out with a separate PR. What do you think about, @mehdi-goli?